hydradx/model/amm/omnipool_amm.py

```diff
@@ -193,18 +335,52 @@ def add_risk_liquidity(
 
     # Token amounts update
     new_state['R'][i] += delta_R
-    new_agents[LP_id]['r'][i] -= delta_R
+    if LP_id:
+        new_agents[LP_id]['r'][i] -= delta_R
+        if new_agents[LP_id]['r'][i] < 0:
+            # print('Transaction rejected because agent has insufficient funds.')
+            # print(f'agent {LP_id}, asset {new_state["token_list"][i]}, amount {delta_R}')
+            return old_state, old_agents
 
     # Share update
-    new_state['S'][i] *= new_state['R'][i] / old_state['R'][i]
-    new_agents[LP_id]['s'][i] += new_state['S'][i] - old_state['S'][i]
+    if new_state['S']:
+        new_state['S'][i] *= new_state['R'][i] / old_state['R'][i]
+    else:
+        new_state['S'] = 1
+
+    if LP_id:
+        # shares go to provisioning agent
+        new_agents[LP_id]['s'][i] += new_state['S'][i] - old_state['S'][i]
+    else:
+        # shares go to protocol
+        new_state['B'] += new_state['S'][i] - old_state['S'][i]
 
-    # LRNA add
+    # LRNA add (mint)
     delta_Q = price_i(old_state, i) * delta_R
     new_state['Q'][i] += delta_Q
 
+    # L update: LRNA fees to be burned before they will start to accumulate again
+    delta_L = delta_R * old_state['Q'][i]/old_state['R'][i] * old_state['L']/sum(old_state['Q'])
+    new_state['L'] += delta_L
+
+    # T update: TVL soft cap
+    stable_index = new_state['stablecoin_index']
+    delta_t = new_state['Q'][i] * new_state['R'][stable_index]/new_state['Q'][stable_index] - new_state['T'][i]
+    new_state['T'][i] += delta_t
+
+    if 'O' in new_state and new_state['Q'][i] / sum(new_state['Q']) > new_state['O'][i]:
+        # print(f'Transaction rejected because it would exceed the weight cap in pool[{i}].')
+        # print(f'agent {LP_id}, asset {new_state["token_list"][i]}, amount {delta_R}')
+        return old_state, old_agents
```

Comment:
Why would TVL be updated if the liquidity add doesn't happen?

Reply:
The TVL we have at the beginning of the liquidity add is *stale* (swaps change TVL but do not update the variable). In theory, even in a failed liquidity transaction, we could update the TVL to be current (given the failed transaction) since we are doing the calculation anyway (and the user is paying the fee for the TVL update already).

Reply:
That would make sense. In practice it may never happen, if the interface is made to warn you before paying for a transaction that will fail. So maybe doesn't matter either way